File: analyzers/tests/SonarAnalyzer.Test/TestCases/ShouldImplementExportedInterfaces.cs
using System;
using System.ComponentModel.Composition;

namespace Classes
{
    static class Constants
    {
        public const string ContractName = "asdasd";
    }

    interface MyInterface { }

    [Export(typeof(MyInterface))] // Noncompliant {{Implement 'MyInterface' on 'NotExported' or remove this export attribute.}}
//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    [Export(typeof(Exported))] // Noncompliant {{Derive from 'Exported' on 'NotExported' or remove this export attribute.}}
//   ^^^^^^^^^^^^^^^^^^^^^^^^
    class NotExported
    {
    }

    [Export(contractType: typeof(IComparable), contractName: "asdasd")] // Noncompliant
    [Export(contractType: typeof(MyInterface), contractName: "asdasd")]
    class NotExported_NamedArgs_ReverseOrder : MyInterface
    {
    }

    [Export("something", typeof(MyInterface))] // Noncompliant
    [Export(Constants.ContractName, typeof(IDisposable))] // Noncompliant
    class NotExported_MultipleArgs
    {
    }

    [Export(typeof(MyInterface)), Export(typeof(IComparable)), Export(typeof(IDisposable))]
//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ {{Implement 'MyInterface' on 'NotExported_Multiple' or remove this export attribute.}}
//                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-1 {{Implement 'IComparable' on 'NotExported_Multiple' or remove this export attribute.}}
    class NotExported_Multiple : IDisposable
    {
        public void Dispose() { }
    }

    [ExportAttribute(typeof(MyInterface))] // Noncompliant
    class NotExported_FullAttributeName
    {
    }

    [Export(typeof(MyInterface))]
    [Export(typeof(Descendant))] // Noncompliant
    class Exported : MyInterface
    {
    }

    [Export]
    [Export("something")] // Exposing ourselves
    [Export(typeof(Exporting_Ourselves))]
    class Exporting_Ourselves
    {
    }

    [Export()]
    class Exporting_CornerCase
    {
    }

    [Export(1)] // Error [CS1503]
    [Export(1, typeof(IComparable))] // Error [CS1503]
    [Export(typeof(ASDASD))] // Error [CS0246]
    [Export(typeof(MyInterface), typeof(IComparable))] // Error [CS1503]
    class InvalidSyntax
    {
    }

    [Import(typeof(MyInterface))] // Error [CS0592] - cannot import here
    [InheritedExport(typeof(MyInterface))] // Noncompliant
    [InheritedExport(typeof(OtherAttributes))]
    class OtherAttributes
    {
    }

    [Export(typeof(MyInterface))]
    [Export(typeof(Exported))]
    class Descendant : Exported
    {
    }

    class ExportingMembers_Are_Ignored
    {
        [Export(typeof(MyInterface))]
        [Export(typeof(Exported))]
        public NotExported MyProperty { get; set; }

        [Export(typeof(MyInterface))]
        [Export(typeof(Exported))]
        public NotExported MyField;

        [Export(typeof(MyInterface))]
        [Export(typeof(Exported))]
        public NotExported MyMethod() { return null; }
    }

    interface ISomething<T> { }
    public class BaseThing { }
    public class BaseThing2 { }

    [Export(typeof(ISomething<BaseThing>))]
    public class BaseSomethingImplementation : ISomething<BaseThing>
    {

    }

    // Error@+1 [CS0416]
    [Export(typeof(ISomething<BaseThing>))] // Noncompliant {{Implement 'ISomething<BaseThing>' on 'Something<BaseThing>' or remove this export attribute.}}
    public class Something<BaseThing>
    {
    }

    [Export(typeof(ISomething<BaseThing>))] // Noncompliant
    public class SomethingImplementation : ISomething<BaseThing2>
    {

    }

    [Export(typeof(ISomething<>))]
    public class Soomething<T> : ISomething<T>
    {
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ShouldImplementExportedInterfaces.vb
Imports System
Imports System.ComponentModel.Composition

Namespace Classes
    Module Constants
        Public Const ContractName As String = "asdasd"
    End Module

    Interface MyInterface
    End Interface

    ' Noncompliant@+2 {{Implement 'MyInterface' on 'NotExported' or remove this export attribute.}}
    ' Noncompliant@+2 {{Derive from 'Exported' on 'NotExported' or remove this export attribute.}}
    <Export(GetType(MyInterface))>
    <Export(GetType(Exported))>
    Class NotExported
'    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-2
'    ^^^^^^^^^^^^^^^^^^^^^^^^^ @-2
    End Class

    ' Noncompliant@+2
    ' Noncompliant@+2
    <Export("something", GetType(MyInterface))>
    <Export(Constants.ContractName, GetType(IDisposable))>
    Class NotExported_MultipleArgs
    End Class

    <Export(GetType(MyInterface)), Export(GetType(IComparable)), Export(GetType(IDisposable))>
    Class NotExported_Multiple
'    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-1 {{Implement 'MyInterface' on 'NotExported_Multiple' or remove this export attribute.}}
'                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-2 {{Implement 'IComparable' on 'NotExported_Multiple' or remove this export attribute.}}
        Implements IDisposable ' Error [BC30149]
        Sub Dispose()
        End Sub
    End Class

    ' Noncompliant@+1
    <ExportAttribute(GetType(MyInterface))>
    Class NotExported_FullAttributeName
    End Class

    ' Noncompliant@+2
    <Export(GetType(MyInterface))>
    <Export(GetType(Descendant))>
    Class Exported
        Implements MyInterface
    End Class

    ' Exposing ourselves
    <Export>
    <Export("something")>
    <Export(GetType(Exporting_Ourselves))>
    Class Exporting_Ourselves
    End Class

    ' Noncompliant@+1
    <InheritedExport(GetType(MyInterface))>
    <InheritedExport(GetType(OtherAttributes))>
    Class OtherAttributes
    End Class

    <Export(GetType(MyInterface))>
    <Export(GetType(Exported))>
    class Descendant
        Inherits Exported
    End Class

    Class ExportingMembers_Are_Ignored
        <Export(GetType(MyInterface))>
        <Export(GetType(Exported))>
        Public Property MyProperty As NotExported

        <Export(GetType(MyInterface))>
        <Export(GetType(Exported))>
        Public MyField As NotExported

        <Export(GetType(MyInterface))>
        <Export(GetType(Exported))>
        Public Function MyMethod() As NotExported
        End Function
    End Class

    Interface ISomething(Of T)
    End Interface

    Public Class BaseThing
    End Class
    Public class BaseThing2
    End Class

    <Export(GetType(ISomething(Of BaseThing)))>
    Public Class BaseSomethingImplementation
        Implements ISomething(Of BaseThing)
    End Class

    ' Noncompliant@+1 {{Implement 'ISomething(Of BaseThing)' on 'Something(Of BaseThing)' or remove this export attribute.}}
    <Export(GetType(ISomething(Of BaseThing)))>
    Public Class Something(Of BaseThing)
    End Class

    ' Noncompliant@+1
    <Export(GetType(ISomething(Of BaseThing)))>
    Public Class SomethingImplementation
        Implements ISomething(Of BaseThing2)
    End Class

    <Export(GetType(ISomething(Of)))>
    Public Class OtherSomething(Of T)
        Implements ISomething(Of T)
    End Class

    ' Noncompliant@+2
    ' Error@+1 [BC31501]
    <Export(ContractType:=GetType(IDisposable))>
    Class Exporting_InvalidSyntax
    End Class
End Namespace
